File: android/app/src/main/kotlin/com/thecityandthebike/ui/viewmodel/ImageDetailViewModel.kt
package com.thecityandthebike.ui.viewmodel

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.thecityandthebike.data.local.TokenManager
import com.thecityandthebike.data.model.ApiResult
import com.thecityandthebike.data.model.dto.SubmissionResponse
import com.thecityandthebike.data.model.dto.TagResponse
import com.thecityandthebike.data.processing.MaskProcessor
import com.thecityandthebike.data.repository.SubmissionRepository
import com.thecityandthebike.data.repository.TagRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ImageDetailState(
    val isLoading: Boolean = false,
    val submission: SubmissionResponse? = null,
    val error: String? = null,
    val isOwner: Boolean = false,
    val isDeleting: Boolean = false,
    val isDeleted: Boolean = false,
    val tags: List<TagResponse> = emptyList(),
    val isTagMode: Boolean = false,
    val isCreatingTag: Boolean = false,
    val isDeletingTag: Boolean = false,
    val isProcessingMask: Boolean = false,
    val processedRing: List<List<Float>>? = null,
    val processedMaskWidth: Int = 0,
    val processedMaskHeight: Int = 0,
)

@HiltViewModel
class ImageDetailViewModel @Inject constructor(
    private val submissionRepository: SubmissionRepository,
    private val tagRepository: TagRepository,
    private val tokenManager: TokenManager,
    private val maskProcessor: MaskProcessor,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    val submissionId: String = checkNotNull(savedStateHandle["submissionId"])

    private val _state = MutableStateFlow(ImageDetailState(isLoading = true))
    val state: StateFlow<ImageDetailState> = _state.asStateFlow()
    private var processMaskJob: Job? = null

    init {
        loadSubmission()
        loadTags()
    }

    private fun loadSubmission() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, error = null)
            when (val result = submissionRepository.getSubmission(submissionId)) {
                is ApiResult.Success -> {
                    val submission = result.data
                    val isOwner = checkOwnership(submission.userId)
                    _state.value = _state.value.copy(
                        isLoading = false,
                        submission = submission,
                        isOwner = isOwner
                    )
                }
                is ApiResult.Error -> {
                    _state.value = _state.value.copy(
                        isLoading = false,
                        error = result.error.displayMessage
                    )
                }
            }
        }
    }

    private fun checkOwnership(submissionUserId: String): Boolean {
        return tokenManager.getUserId() == submissionUserId
    }

    fun deleteSubmission() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isDeleting = true)
            when (submissionRepository.deleteSubmission(submissionId)) {
                is ApiResult.Success -> {
                    _state.value = _state.value.copy(isDeleting = false, isDeleted = true)
                }
                is ApiResult.Error -> {
                    _state.value = _state.value.copy(
                        isDeleting = false,
                        error = "Failed to delete submission"
                    )
                }
            }
        }
    }

    fun clearError() {
        _state.value = _state.value.copy(error = null)
    }

    private fun loadTags() {
        viewModelScope.launch {
            when (val result = tagRepository.getTags(submissionId)) {
                is ApiResult.Success -> {
                    _state.value = _state.value.copy(tags = result.data)
                }
                is ApiResult.Error -> {
                    // Silently fail - tags are supplementary
                }
            }
        }
    }

    fun enterTagMode() {
        _state.value = _state.value.copy(isTagMode = true)
    }

    fun exitTagMode() {
        processMaskJob?.cancel()
        processMaskJob = null
        _state.value = _state.value.copy(
            isTagMode = false,
            processedRing = null,
            processedMaskWidth = 0,
            processedMaskHeight = 0
        )
    }

    fun processMask(imageFile: java.io.File) {
        processMaskJob = viewModelScope.launch {
            _state.value = _state.value.copy(isProcessingMask = true)
            try {
                val result = maskProcessor.process(imageFile)
                _state.value = _state.value.copy(
                    isProcessingMask = false,
                    processedRing = result.ring,
                    processedMaskWidth = result.width,
                    processedMaskHeight = result.height
                )
            } catch (_: IllegalArgumentException) {
                _state.value = _state.value.copy(
                    isProcessingMask = false,
                    error = "Failed to process mask"
                )
            }
        }
    }

    fun goBackToDrawing() {
        _state.value = _state.value.copy(
            processedRing = null,
            processedMaskWidth = 0,
            processedMaskHeight = 0
        )
    }

    fun createTag(imageFile: java.io.File) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isCreatingTag = true)
            val currentState = _state.value
            when (val result = tagRepository.createTag(
                submissionId,
                imageFile,
                ring = currentState.processedRing,
                ringWidth = currentState.processedMaskWidth.takeIf { it > 0 },
                ringHeight = currentState.processedMaskHeight.takeIf { it > 0 }
            )) {
                is ApiResult.Success -> {
                    _state.value = _state.value.copy(
                        isCreatingTag = false,
                        isTagMode = false,
                        processedRing = null,
                        processedMaskWidth = 0,
                        processedMaskHeight = 0,
                        tags = listOf(result.data) + _state.value.tags
                    )
                }
                is ApiResult.Error -> {
                    _state.value = _state.value.copy(
                        isCreatingTag = false,
                        error = "Failed to create tag"
                    )
                }
            }
        }
    }

    fun deleteTag(tagId: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isDeletingTag = true)
            when (tagRepository.deleteTag(tagId)) {
                is ApiResult.Success -> {
                    _state.value = _state.value.copy(
                        isDeletingTag = false,
                        tags = _state.value.tags.filter { it.tagId != tagId }
                    )
                }
                is ApiResult.Error -> {
                    _state.value = _state.value.copy(
                        isDeletingTag = false,
                        error = "Failed to delete tag"
                    )
                }
            }
        }
    }

    fun isLoggedIn(): Boolean {
        return tokenManager.getUserId() != null
    }

    fun isTagOwner(tag: TagResponse): Boolean {
        return tokenManager.getUserId() == tag.userId
    }
}

File: ios/Packages/TCATBFeed/Sources/TCATBFeed/ImageDetailViewModel.swift
import Foundation
import Observation
import TCATBModels

@MainActor
@Observable
public final class ImageDetailViewModel {
    public private(set) var isLoading = false
    public private(set) var submission: SubmissionResponse?
    public private(set) var error: String?
    public private(set) var isOwner = false
    public private(set) var isDeleting = false
    public private(set) var isDeleted = false
    public private(set) var tags: [TagResponse] = []
    public var selectedTagId: String?
    public private(set) var isTagMode = false
    public private(set) var isCreatingTag = false
    public private(set) var isDeletingTag = false
    public private(set) var isProcessingMask = false
    public private(set) var processedRing: [[Float]]?
    public private(set) var processedMaskWidth: Int = 0
    public private(set) var processedMaskHeight: Int = 0
    public private(set) var pointsAwarded: [ScoringBreakdown]?
    public private(set) var isFlagged = false
    public private(set) var isFlagging = false
    public private(set) var flagCount: Int = 0
    public private(set) var isAdmin = false
    public private(set) var isClearingFlags = false

    public let submissionId: String
    private let apiClient: FeedAPIClient
    private let currentUserId: String?

    public init(submissionId: String, apiClient: FeedAPIClient, currentUserId: String?, isAdmin: Bool = false) {
        self.submissionId = submissionId
        self.apiClient = apiClient
        self.currentUserId = currentUserId
        self.isAdmin = isAdmin
    }

    public var isLoggedIn: Bool {
        currentUserId != nil
    }

    public func load() async {
        isLoading = true
        error = nil

        async let submissionTask: Void = loadSubmission()
        async let tagsTask: Void = loadTags()
        async let flagsTask: Void = loadFlagStatus()

        _ = await (submissionTask, tagsTask, flagsTask)

        isLoading = false
    }

    private func loadSubmission() async {
        do {
            let result = try await apiClient.getSubmission(id: submissionId)
            submission = result
            isOwner = result.userId == currentUserId
        } catch {
            self.error = error.localizedDescription
        }
    }

    private func loadTags() async {
        do {
            tags = try await apiClient.getTags(submissionId: submissionId)
        } catch {
            // Tags are supplementary; silently fail
        }
    }

    private func loadFlagStatus() async {
        guard currentUserId != nil else { return }
        do {
            let result = try await apiClient.getFlagStatus(submissionId: submissionId)
            isFlagged = result.flagged
            flagCount = result.flagCount
        } catch {
            // Flag status is supplementary; silently fail
        }
    }

    public func deleteSubmission() async {
        isDeleting = true
        do {
            try await apiClient.deleteSubmission(id: submissionId)
            isDeleted = true
        } catch {
            self.error = "Failed to delete submission"
        }
        isDeleting = false
    }

    public func clearError() {
        error = nil
    }

    // MARK: - Tag Mode

    public func enterTagMode() {
        isTagMode = true
    }

    public func exitTagMode() {
        isTagMode = false
        processedRing = nil
        processedMaskWidth = 0
        processedMaskHeight = 0
    }

    public func processMask(imageData: Data) async {
        isProcessingMask = true
        do {
            let result = try await apiClient.processMask(imageData: imageData)
            processedRing = result.ring
            processedMaskWidth = result.width
            processedMaskHeight = result.height
        } catch {
            self.error = "Failed to process mask"
        }
        isProcessingMask = false
    }

    public func goBackToDrawing() {
        processedRing = nil
        processedMaskWidth = 0
        processedMaskHeight = 0
    }

    public func createTag(imageData: Data) async {
        isCreatingTag = true
        do {
            let tag = try await apiClient.createTag(
                submissionId: submissionId,
                imageData: imageData,
                ring: processedRing,
                ringWidth: processedMaskWidth > 0 ? processedMaskWidth : nil,
                ringHeight: processedMaskHeight > 0 ? processedMaskHeight : nil
            )
            let breakdown = tag.pointsBreakdown
            tags.insert(tag, at: 0)
            isTagMode = false
            processedRing = nil
            processedMaskWidth = 0
            processedMaskHeight = 0
            if let breakdown, !breakdown.isEmpty {
                pointsAwarded = breakdown
            }
        } catch {
            self.error = "Failed to create tag"
        }
        isCreatingTag = false
    }

    public func deleteTag(_ tagId: String) async {
        isDeletingTag = true
        do {
            try await apiClient.deleteTag(id: tagId)
            tags.removeAll { $0.tagId == tagId }
            if selectedTagId == tagId {
                selectedTagId = nil
            }
        } catch {
            self.error = "Failed to delete tag"
        }
        isDeletingTag = false
    }

    public func isTagOwner(_ tag: TagResponse) -> Bool {
        currentUserId == tag.userId
    }

    public func clearPointsAwarded() {
        pointsAwarded = nil
    }

    // MARK: - Flagging

    public func flagSubmission() async {
        isFlagging = true
        do {
            let result = try await apiClient.createFlag(submissionId: submissionId)
            isFlagged = result.flagged
            flagCount = result.flagCount
        } catch {
            self.error = "Failed to flag submission"
        }
        isFlagging = false
    }

    public func clearFlags() async {
        isClearingFlags = true
        do {
            let result = try await apiClient.clearFlags(submissionId: submissionId)
            flagCount = result.flagCount
            isFlagged = result.flagged
        } catch {
            self.error = "Failed to clear flags"
        }
        isClearingFlags = false
    }
}
